### DISCUSSION TITLE
Create a first API to navigate events

### DISCUSSION THREAD

**Discussion by alexgarel** [openfoodfacts/openfoodfacts-query#148] [Ideas]
We have an event database and it can be useful in a number of way.

An important usecase is to browse last edits on some specific part of the data.

1. Could we have an API akin to product opener facets and get last edits as JSON data.
   
   going to query.openfoodfacts.org/facets/category/en:beer/label/en:gluten-free would give me last edits on gluten free beers

   This can help someone monitoring events on a certain category to correct data, or see new additions.

2. Could we also have an API to get last events of a certain type made by a particular user ?
   
   This could be used to create a better view of lastly modified / added products for the user on Open Food Facts.

**Comment by LSUDOKO** [comment]
Can you assign me this issue

**Comment by alexgarel** [comment]
@LSUDOKO you are welcome to start working on this, but you have to wait for the[ Python rewrite to be merged](https://github.com/openfoodfacts/openfoodfacts-query/pull/126).

You should also validate your approach with @john-gom (he is the maintainer).

**Comment by john-gom** [comment]
Hi @LSUDOKO it would be great to get you involved. As a start maybe come up with a proposed API specification. It would be good to have something we can use for other types of query too as off-query currently emulates the MongoDB syntax which can be difficult to grasp. It would also be helpful to get your input on the current Python refactor and any experiences you have setting up for local development

**Comment by john-gom** [comment]
Hi @alexgarel . I'm starting to look at this (@LSUDOKO please let me know if you would prefer to do it)...

I think the `query.openfoodfacts.org/facets/category/en:beer/label/en:gluten-free` convention works OK for aggregations, but I'm not sure it is very intuitive for filtering and it could be difficult to document in OpenAPI. I'd suggest filtering using query parameters, e.g.

`query.openfoodfacts.org/events?categories_tags=en:beer&labels_tags=en:gluten-free&limit=50`

This would return specific events. e.g. `[{"code":"1234567","rev":2,"action":"updated","user_id":"alexg","diffs":{...}},{"code":"5876876", ...}]` If we want aggregate counts then maybe something like:

`query.openfoodfacts.org/events/contributors?categories_tags=en:beer&labels_tags=en:gluten-free&from=2024-01-01&to=2025-01-01`

Which would return the counts by contributor in the date range specified, perhaps further grouped by action, e.g. `{"alexg":{"created":23,"updated": 66,"deleted":1},"stephane":{"updated":67},...}`

**Reply by alexgarel** [reply]
Yes, looks good to me.
